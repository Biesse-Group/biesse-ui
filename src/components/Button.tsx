import { FC, PropsWithChildren } from "react";
import styled, { css } from "styled-components";

export interface ButtonProps {
  /**
   * Optional component class name
   */
  className?: string;
  /**
   * Is this the principal call to action on the page?
   */
  variant: "primary" | "outline";
  /**
   * How large should the button be?
   */
  size?: "small" | "medium";
  /**
   * Full-width button
   */
  isBlock?: boolean;
  /**
   * Callback on click
   */
  onClick?: () => void;
}

const getSizeStyle = (size: ButtonProps["size"]) => {
  switch (size) {
    case "small":
      return css`
        font-size: ${(props) => props.theme.font.body.sm};
        padding: 8px 26px;
      `;
    default:
      return css`
        font-size: ${(props) => props.theme.font.body.md};
        padding: 13px 32px;
      `;
  }
};

const getVariantStyle = (variant: ButtonProps["variant"]) => {
  return variant === "primary"
    ? css`
        color: ${(props) => props.theme.color.white};
        background-color: ${(props) => props.theme.color.primary};
        border: 1px solid ${(props) => props.theme.color.primary};

        &:hover {
          color: ${(props) => props.theme.color.primary};
          background-color: ${(props) => props.theme.color.white};
        }
      `
    : css`
        background-color: transparent;
        color: ${(props) => props.theme.color.primary};
        border: 1px solid ${(props) => props.theme.color.primary};

        &:hover {
          background-color: ${(props) => props.theme.color.primary};
          color: ${(props) => props.theme.color.white};
        }
      `;
};

const ButtonStyled = styled.button<ButtonProps>`
  font-family: ${(props) => props.theme.font.family};
  font-weight: bold;
  border: 0;
  border-radius: ${(props) => props.theme.borderRadius.md};
  cursor: pointer;
  display: inline-block;
  line-height: 1;
  transition: all 0.2s ease-out;
  text-transform: uppercase;
  line-height: 1;

  ${(props) => getSizeStyle(props.size)}
  ${(props) => getVariantStyle(props.variant)}

  ${(props) =>
    props.isBlock &&
    css`
      width: 100%;
    `}
`;

export const Button: FC<PropsWithChildren<ButtonProps>> = (props) => {
  return <ButtonStyled {...props} />;
};
