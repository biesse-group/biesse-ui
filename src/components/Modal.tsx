import { AnimatePresence, motion } from "framer-motion";
import { FC, PropsWithChildren } from "react";
import styled, { css } from "styled-components";

import { mqUntil } from "../styles";
import { IconButton } from "./IconButton";

export interface ModalProps {
  /**
   * Optional component class name
   */
  className?: string;
  /**
   * Whether the modal should be hidden or shown
   */
  isShown?: boolean;
  /**
   * Action issued when clicking outside the video or on the dedicated button
   * Should be wired to the isShown state
   */
  onCloseAction: () => void;
  testId?: string;
}

const Overlay = styled(motion.div)`
  width: 100%;
  height: 100%;
  background-color: ${(props) => props.theme.color.modalBackground};
  position: absolute;
  top: 0;
  left: 0;
  z-index: 10000;
  display: flex;
  align-items: center;
  justify-content: center;
  overflow: hidden;
`;

const ChildrenContainer = styled.div`
  z-index: 1;
`;

const StyledCloseButton = styled(IconButton)`
  position: absolute;
  top: 75px;
  right: 90px;
  z-index: 2;

  ${mqUntil(
    "sm",
    css`
      top: 20px;
      right: 20px;
    `
  )}
`;

export const Modal: FC<PropsWithChildren<ModalProps>> = ({
  testId,
  isShown,
  onCloseAction,
  ...props
}) => {
  return (
    <AnimatePresence>
      {isShown && (
        <Overlay
          data-testid={testId}
          onClick={onCloseAction}
          key="modal"
          initial={{ opacity: 0 }}
          animate={{ opacity: 1 }}
          exit={{ opacity: 0 }}
        >
          <StyledCloseButton
            aria-label="close"
            variant="light"
            icon="close"
            onClick={onCloseAction}
          />
          <ChildrenContainer {...props} onClick={(e) => e.stopPropagation()}></ChildrenContainer>
        </Overlay>
      )}
    </AnimatePresence>
  );
};
