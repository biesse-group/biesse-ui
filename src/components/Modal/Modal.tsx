import {
  FloatingFocusManager,
  FloatingOverlay,
  FloatingPortal,
  useClick,
  useDismiss,
  useFloating,
  useId,
  useInteractions,
  useRole,
} from "@floating-ui/react";
import { AnimatePresence, motion } from "framer-motion";
import type { FC, PropsWithChildren } from "react";
import styled, { css } from "styled-components";

import type { BaseProps } from "~components/baseProps";
import { IconButton } from "~components/IconButton";
import { mqUntil } from "~styles";

export interface ModalProps extends BaseProps {
  testId?: string;
  isOpen: boolean;
  close: () => void;
}

const StyledOverlay = styled(FloatingOverlay)`
  background-color: ${(props) => props.theme.color.modalBackground};
  display: flex;
  z-index: 10000;
  overflow: hidden;
`;

const StyledCloseButton = styled(IconButton)`
  position: absolute;
  top: 75px;
  right: 90px;
  z-index: 2;

  ${mqUntil(
    "sm",
    css`
      top: 20px;
      right: 20px;
    `
  )}
`;

const ModalWrapper = styled.div`
  width: 100vw;
  height: 100vh;
  overflow-y: auto;
  position: fixed;
  z-index: 100000;
`;

export const Modal: FC<PropsWithChildren<ModalProps>> = ({
  testId,
  children,
  isOpen,
  close,
  ...props
}) => {
  const { refs, context } = useFloating({
    open: isOpen,
    onOpenChange: close,
  });

  const click = useClick(context);
  const role = useRole(context);
  const dismiss = useDismiss(context, { outsidePressEvent: "mousedown" });

  const { getFloatingProps } = useInteractions([click, role, dismiss]);

  const headingId = useId();
  const descriptionId = useId();

  return (
    <FloatingPortal>
      <AnimatePresence>
        {isOpen && (
          <motion.div
            key="modal"
            initial={{ opacity: 0 }}
            animate={{ opacity: 1 }}
            exit={{ opacity: 0 }}
            {...props}
          >
            <StyledOverlay className="Dialog-overlay" lockScroll>
              <FloatingFocusManager context={context}>
                <ModalWrapper
                  ref={refs.setFloating}
                  data-testid={testId}
                  aria-labelledby={headingId}
                  aria-describedby={descriptionId}
                  {...getFloatingProps()}
                >
                  <StyledCloseButton
                    aria-label="close"
                    variant="primary-inverted"
                    icon="close"
                    onClick={close}
                  />
                  {children}
                </ModalWrapper>
              </FloatingFocusManager>
            </StyledOverlay>
          </motion.div>
        )}
      </AnimatePresence>
    </FloatingPortal>
  );
};
