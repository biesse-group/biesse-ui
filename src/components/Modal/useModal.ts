import { useCallback, useEffect, useRef, useState } from "react";

export interface UseModalProps {
  defaultIsOpen?: boolean;
  onClose?: () => void;
  onOpen?: () => void;
}

export function useModal(props?: UseModalProps) {
  const { onOpen, onClose, defaultIsOpen } = props ?? {};
  const [isOpen, setIsOpen] = useState(Boolean(defaultIsOpen));
  const wasOpen = useRef(isOpen);

  const open = useCallback(() => {
    setIsOpen(true);
    wasOpen.current = true;
  }, []);
  const close = useCallback(() => setIsOpen(false), []);

  useEffect(() => {
    if (onOpen && isOpen) {
      onOpen();
    } else if (onClose && !isOpen && wasOpen.current) {
      onClose();
    }
  }, [onOpen, onClose, isOpen]);

  return {
    isOpen,
    open,
    close,
  };
}
