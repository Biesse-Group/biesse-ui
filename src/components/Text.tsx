import { FC, PropsWithChildren } from "react";
import styled, { css } from "styled-components";

import { BiesseTheme } from "../themes";

export type TextProps = {
  size?: keyof BiesseTheme["font"]["body"];
  weight?: keyof BiesseTheme["font"]["weight"];
  color?: "default" | "primary" | "light";
};

const getColor = (color?: TextProps["color"]) => css`
  color: ${(props) => {
    switch (color) {
      case "light":
        return props.theme.color.white;
      case "primary":
        return props.theme.color.primary;
      default:
        return props.theme.color.black;
    }
  }};
`;

const getLineHeight = (size?: TextProps["size"]) => css`
  line-height: ${(props) => {
    switch (size) {
      case "xs":
        return "18px";
      case "sm":
        return "20px";
      case "lg":
        return "26px";
      default:
        return "24px";
    }
  }};
`;

const StyledText = styled.span<TextProps>`
  font-family: ${(props) => props.theme.font.family};
  font-weight: ${(props) => props.theme.font.weight[props.weight || "book"]};
  font-size: ${(props) => props.theme.font.body[props.size || "md"]};
  ${(props) => getLineHeight(props.size)};
  ${(props) => getColor(props.color)};
`;

export const Text: FC<PropsWithChildren<TextProps>> = (props) => {
  return <StyledText {...props} />;
};
