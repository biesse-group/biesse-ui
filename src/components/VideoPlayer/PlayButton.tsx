import { AnimatePresence, motion } from "framer-motion";
import { FC } from "react";
import styled from "styled-components";

import { Icon } from "../Icon";

export interface PlayButtonProps {
  isPlaying?: boolean;
  onClick?: () => void;
  testId?: string;
}

const PlayButtonRoot = styled(motion.div)`
  position: relative;
  display: flex;
  height: 80px;
  width: 80px;
  align-items: center;
  justify-content: center;
  border: 3px solid ${(props) => props.theme.color.white};
  border-radius: 50px;
  background-color: ${(props) => props.theme.color.primary};
  box-shadow: ${(props) => props.theme.videoPlayer.buttonBoxShadow};
  cursor: pointer;
  overflow: hidden;
`;

const InternalBorder = styled.div`
  height: 100%;
  width: 100%;
  border: 3px solid ${(props) => props.theme.color.white};
  opacity: 0.4;
  border-radius: 50px;
  z-index: 1;

  position: absolute;
`;

const StyledIcon = styled(Icon)`
  margin-left: 7px;
`;

export const PlayButton: FC<PlayButtonProps> = ({ testId, isPlaying, ...props }) => {
  return (
    <AnimatePresence>
      {!isPlaying && (
        <PlayButtonRoot
          key="modal"
          initial={{ opacity: 0 }}
          animate={{ opacity: 1 }}
          exit={{ opacity: 0 }}
          data-testid={testId}
          {...props}
        >
          <InternalBorder />
          <StyledIcon name="play" color="light" size="30px" />
        </PlayButtonRoot>
      )}
    </AnimatePresence>
  );
};
