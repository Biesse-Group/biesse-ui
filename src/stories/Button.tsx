/** @jsxImportSource @emotion/react */
import { css } from "@emotion/react";
import React, { FC } from "react";

const buttonStyle = css`
  font-family: "Nunito Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
  font-weight: 700;
  border: 0;
  border-radius: 3em;
  cursor: pointer;
  display: inline-block;
  line-height: 1;
  font-size: 14px;
  padding: 11px 20px;
`;

const buttonPrimary = css`
  color: white;
  background-color: #1ea7fd;
`;

const buttonSecondary = css`
  color: #333;
  background-color: transparent;
  box-shadow: rgba(0, 0, 0, 0.15) 0px 0px 0px 1px inset;
`;

const buttonSmall = css`
  font-size: 12px;
  padding: 10px 16px;
`;

const buttonLarge = css`
  font-size: 16px;
  padding: 12px 24px;
`;

interface ButtonProps {
  /**
   * Is this the principal call to action on the page?
   */
  primary?: boolean;
  /**
   * How large should the button be?
   */
  size?: "small" | "medium" | "large";
  /**
   * Button contents
   */
  label: string;
  /**
   * Optional click handler
   */
  onClick?: () => void;
}

/**
 * Primary UI component for user interaction
 */
export const Button: FC<ButtonProps> = ({ primary = false, size = "medium", label, ...props }) => {
  return (
    <button
      type="button"
      css={[
        buttonStyle,
        primary ? buttonPrimary : buttonSecondary,
        size === "large" && buttonLarge,
        size === "small" && buttonSmall,
      ]}
      {...props}
    >
      {label}
    </button>
  );
};
