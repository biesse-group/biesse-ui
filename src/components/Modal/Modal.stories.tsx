import { type Meta, type StoryObj } from "@storybook/react";
import { type FC, type PropsWithChildren } from "react";

import { Button } from "~components/Button";
import { Text } from "~components/Text";
import { VideoPlayer } from "~components/VideoPlayer";

import { Modal, type ModalProps } from "./Modal";
import { useModal } from "./useModal";

const ModalContainer: FC<
  PropsWithChildren<ModalProps> & {
    label?: string;
    onClose?: () => void;
    onOpen?: () => void;
  }
> = ({ children, label, onClose, onOpen }) => {
  const { isOpen, open, close } = useModal({ onOpen, onClose });
  return (
    <div style={{ maxWidth: 500 }}>
      <Button variant="primary" onClick={open}>
        {label ?? "Open modal"}
      </Button>
      <Modal isOpen={isOpen} close={close}>
        {children}
      </Modal>
    </div>
  );
};

const meta: Meta<typeof ModalContainer> = {
  title: "Dialogs/Modal",
  component: ModalContainer,
  tags: ["autodocs"],
  argTypes: {
    children: {
      control: false,
    },
  },
};

export default meta;

type Story = StoryObj<typeof ModalContainer>;

export const Default: Story = {
  args: {
    children: (
      <div
        style={{
          height: "100%",
          width: "100%",
          display: "flex",
          justifyContent: "center",
          alignItems: "center",
        }}
      >
        <Text color="light">I am the modal content</Text>
      </div>
    ),
  },
};

export const WithVideo: Story = {
  args: {
    label: "Play video",
    children: (
      <VideoPlayer
        variant="fit"
        autoPlay={true}
        url="http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/ElephantsDream.mp4"
      />
    ),
  },
};
