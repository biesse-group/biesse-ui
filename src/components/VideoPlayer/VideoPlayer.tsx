import { createRef, FC, PropsWithChildren, useState } from "react";
import styled, { css } from "styled-components";

import { mqUntil } from "../../styles";
import { PlayButton } from "./PlayButton";

export interface VideoPlayerProps {
  /**
   * Optional component class name
   */
  className?: string;
  url: string;
  /**
   * Video format, if none is passed mp4 is assumed
   */
  mimeType?: string;
  loop?: boolean;
  testId?: string;
}

const VideoPlayerRoot = styled.div`
  position: relative;
  overflow: hidden;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
`;

const PlayButtonWrapper = styled.div`
  position: absolute;
  height: 100%;
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const PauseAction = styled.div<{ isPlaying: boolean }>`
  position: absolute;
  width: 100%;
  height: 100%;
  z-index: 1;
  cursor: pointer;

  ${(props) =>
    !props.isPlaying &&
    css`
      display: none;
    `}
`;

const StyledVideo = styled.video`
  width: 100%;

  ${mqUntil(
    "sm",
    css`
      width: auto;
      height: 100%;
    `
  )}
`;

export const VideoPlayer: FC<PropsWithChildren<VideoPlayerProps>> = ({
  testId,
  url,
  mimeType,
  loop = true,
  ...props
}) => {
  const videoRef = createRef<HTMLVideoElement>();

  const [isVideoPlaying, setIsVideoPlaying] = useState<boolean>(false);

  const handlePlayButton = () => {
    if (!isVideoPlaying) {
      videoRef.current?.play();
    }
  };
  const handlePause = () => {
    if (isVideoPlaying) {
      videoRef.current?.pause();
    }
  };

  return (
    <VideoPlayerRoot data-testid={testId} {...props}>
      <PauseAction isPlaying={isVideoPlaying} onClick={handlePause} data-testid={`pause-action`} />
      <StyledVideo
        onPlaying={() => setIsVideoPlaying(true)}
        onPause={() => setIsVideoPlaying(false)}
        ref={videoRef}
        loop={loop}
      >
        <source src={url} type={mimeType || "video/mp4"} />
        Your browser does not support HTML video.
      </StyledVideo>

      <PlayButtonWrapper>
        <PlayButton
          testId={testId ? `play-button` : undefined}
          isPlaying={isVideoPlaying}
          onClick={handlePlayButton}
        />
      </PlayButtonWrapper>
    </VideoPlayerRoot>
  );
};
