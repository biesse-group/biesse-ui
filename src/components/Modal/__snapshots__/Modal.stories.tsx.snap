// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialogs/Modal Default smoke-test 1`] = `
<div style="max-width: 500px;">
  <button type="button"
          class="sc-fqkvVR bxTuzs"
  >
    Open modal
  </button>
</div>
`;

exports[`Dialogs/Modal WithVideo smoke-test 1`] = `
<div style="max-width: 500px;">
  <button type="button"
          class="sc-fqkvVR bxTuzs"
  >
    Play video
  </button>
</div>
`;
